rules/cc_side_outputs.bzl

```diff
@@ -0,0 +1,281 @@
+# Copyright lowRISC contributors.
+# Licensed under the Apache License, Version 2.0, see LICENSE for details.
+# SPDX-License-Identifier: Apache-2.0
+
+"""Aspects and rules for making cc_* libraries emit more outputs."""
+
+load("@rules_cc//cc:action_names.bzl", "ACTION_NAMES", "C_COMPILE_ACTION_NAME")
+load("//rules:bugfix.bzl", "find_cc_toolchain")
+load("//rules:rv.bzl", "rv_rule")
+load("//rules:util.bzl", "expand_tabs_in_artifact")
+
+CcSideProductInfo = provider(fields = ["files"])
+
+def _cc_compile_different_output(target, ctx, extension, flags):
+    """
+    Helper macro for implementing the .s and .ll outputting libraries.
+
+    In addition to the usual target and ctx inputs for an aspect, it also
+    takes an extension to add to each source file of the rule being analyzed,
+    and flags to add to the compiler arguments to get the output we want.
+    """
+    if ctx.rule.kind not in ["cc_library", "cc_binary", "cc_test"]:
+        return CcSideProductInfo(files = [])
+
+    # This filters out both headers and assembly source inputs, neither of which
+    # make sense to generate an .s output for.
+    translation_units = [
+        src.files.to_list()[0]
+        for src in ctx.rule.attr.srcs
+        if src.files.to_list()[0].extension in [
+            # We only use .cc, but to avoid nasty surprises we support all five
+            # C++ file extensions.
+            "c",
+            "cc",
+            "cpp",
+            "cxx",
+            "c++",
+            "C",
+        ]
+    ]
+
+    # Libraries consisting of just headers or assembly files have nothing
+    # useful to contribute to the output.
+    if len(translation_units) == 0:
+        return CcSideProductInfo(files = [])
+
+    if CcInfo in target:
+        cc_info = target[CcInfo]
+    else:
+        # Some rules, like cc_binary, do NOT produce a CcInfo provider. Therefore,
+        # we need to build one from its dependencies.
+        cc_info = cc_common.merge_cc_infos(
+            direct_cc_infos = [dep[CcInfo] for dep in ctx.rule.attr.deps if CcInfo in dep],
+        )
+    cc_compile_ctx = cc_info.compilation_context
+
+    cc_toolchain = find_cc_toolchain(ctx)
+    feature_configuration = cc_common.configure_features(
+        ctx = ctx,
+        cc_toolchain = cc_toolchain,
+        requested_features = ctx.features,
+        unsupported_features = ctx.disabled_features,
+    )
+
+    c_compiler_path = cc_common.get_tool_for_action(
+        feature_configuration = feature_configuration,
+        action_name = ACTION_NAMES.c_compile,
```

Comment:
There is in theory a different compiler used for C and for C++, so to be perfectly safe you could `get_tool_for_action` also for `cpp_compile` action.

Reply:
I don't think that's a thing we actually need to worry about in this project; we don't support anything other than GCC or Clang (soon to be Clang only, in some cases!).